### ISSUE TITLE
Improve makefile targets

### ISSUE THREAD

**Issue by ghost** [ocaml/v3.ocaml.org-rescript#202]
Let's implement makefile version 4.

- [x] Remove nvm and npx from makefiles (nvm and npx make the commands too verbose, and mandating them makes too many assumptions. for example, CI images won't use nvm.)
- [x] Add yarn to engines in package.json
- [x] Update readme to recommend using nvm, but do not require it
- [x] Perform yarn install and esy install and build in install-deps. When install-deps is complete, all ocaml utilities should be built and ready to invoke.
- [x] Remove comments in ci* makefile targets about installing nvm
- [x] Update contributing instructions to use yarn scripts directly

Create follow-up issue to define a check prerequisites target which ensures that acceptable versions of node, esy, and yarn are installed.

**Comment by ghost** [comment]
Some documentation on the eventual layers within the build infrastructure is available [here](../wikis/Ecosystem-Release-Notes#dependency-layers).


TODO: change the list into a directed acyclic graph

**Comment by ghost** [comment]
After this issue is done, we are assuming its the developers responsibility to read up on nvm and to upgrade esy or yarn if their machine has stale versions of these utilities.

**Comment by ghost** [comment]
Should we perform global installation of yarn and esy in install-deps, or should we only provide instructions or an additional optional target for installing those global tools?

**Comment by agarwal** [comment]
> Should we perform global installation of yarn and esy in install-deps

My opinion is no. All dependencies should be local to the project if at all possible.

**Comment by ghost** [comment]
> All dependencies should be local to the project if at all possible

We should synchronize on this point. The current direction of this issue is to move from a local install of yarn to a global install of yarn to reduce the use of npx.

From limited experiments with esy, we have encountered problems trying to install esy locally. The question above was pondering whether we should perform the global installation or just suggest the action to the user.

**Comment by ghost** [comment]
>  All dependencies should be local to the project if at all possible.

This is how the project is currently setup (besides the upcoming addition of esy). It has operated pretty smoothly, but I have noticed minor drawbacks.

* The commands get long and often require a makefile target wrapping a yarn script. This also introduces friction that slows down developers from quickly adding more yarn scripts.
* The CI and local development commands drift a little, as they are different and need to be maintained separately. CI systems usually have tools globally installed and don't use nvm.
* ~~You lose auto completion for tools like yarn. `npx yarn` doesn't offer autocompletion of script names for me.~~ We do not know if autocompletion is impacted by using `npx`. We might research this topic later on.
* ~~When you install yarn locally, it pollutes and could rarely even clash with the transitive dependencies of other libraries being used within the project. If `yarn` and `react` clashed on allowed version of a library, that could slow down an upgrade.~~ Irrelevant, see comment below.
* Ideally, a developer should be fluent in all the subcommands of yarn, bsb, and next. Since we have a layer around yarn and other tools, one is less likely to use and be familiar with the underyling tools. Another small benefit of using the underlying commands is that developers will become accustomed to consult the yarn, bsb, and next command help when they are unsure about the behavior in corner cases.

The biggest advantage of having all tools locally installed is that users transparently upgrade tools like yarn, but this happens only once every six months or year.

**Comment by ghost** [comment]
> *    Remove nvm and npx from makefiles (nvm and npx make the commands too verbose, and mandating them makes too many assumptions. for example, CI images won't use nvm.)
> *    Add yarn to engines in package.json
> *    Update readme to recommend using nvm, but do not require it
> *    Perform yarn install and esy install and build in install-deps. When install-deps is complete, all ocaml utilities should be built and ready to invoke.
> *    Remove comments in ci* makefile targets about installing nvm
> *    Update contributing instructions to use yarn scripts directly

Above were completed in https://github.com/ocaml/v3.ocaml.org/pull/301 .

This issue will remain open until we have time to discuss the desired outcome.

**Comment by agarwal** [comment]
- [x] Be consistent with required version of `node`. The `package.json` file says `14.x`, so any version in that range is allowed. That also matches what we select in Vercel (which does not allow more specific versions to be selected). However, our `.nvmrc` file says `14.15.1`, implying inconsistently that we require that very specific version.

**Comment by ghost** [comment]
Update: Kanishka and Ashish met on 4/26 and decided on a new direction. The new target names will be: install-deps, watch, build, serve, ci-install-deps, and ci-build.

We found that installing `yarn` globally on MacOS takes too much online research. We also are willing to sacrifice command reuse between CI (which doesn't use `nvm`) and local development (which may use `nvm`) in exchange for more automation for local development. We are reverting all the local development commands to make use of `nvm`. We will have all tools installed locally when possible, including `yarn`. We will continue to request that the user install `esy` globally, since we haven't found a way to properly install `esy` locally in all settings. Regardless of whether a tool like `yarn` or `esy` is actually installed locally or globally, we will use the indirection of using `npx TOOL@version` to allow the user the flexibility of installing a tool globally if they wish to do so.

**Comment by ghost** [comment]
- [x] Make the `engine` version range for `yarn` more narrow.

**Comment by agarwal** [comment]
> could rarely even clash with the transitive dependencies of other libraries being used within the project

We just checked on https://npmgraph.js.org/ and https://www.npmjs.com/package/yarn, and actually `yarn` has no dependencies. So this point is moot. (Though we're not sure how since their [package.json](https://github.com/yarnpkg/yarn/blob/master/package.json) does list a lot of dependencies.)

**Comment by ghost** [comment]
Here is the thread where the ReScript team explains that there is a newer option for enabling coloring: https://forum.rescript-lang.org/t/nextjs-concurrently-and-terminal-coloring/1532 .

**Comment by agarwal** [comment]
@rdavison I see you decided to include `install-deps` in the default `Makefile` target. I'm double minded about this. That step takes 13 seconds on my reasonably fast computer, when there are no changes to make. Thus, I might get annoyed doing that every time I start building this project. I'll probably end up typing `make watch-and-serve`, i.e. I'll always not use the default target.

The benefit of your decision is that people who infrequently develop will be less likely to forget to upgrade the dependencies. But maybe people have to remember to run `make install-reps` every once in a while.

So the question is who is the default target for: i) frequent developers, or ii) new or infrequent developers?

**Comment by rdavison** [comment]
The way I see it is that you only have to run the Makefile once, as it puts you in the watch mode immediately.  I don't mind waiting 13 seconds if I only have to do it once per coding session.

**Comment by rdavison** [comment]
> Be consistent with required version of node. The package.json file says 14.x, so any version in that range is allowed. That also matches what we select in Vercel (which does not allow more specific versions to be selected). However, our .nvmrc file says 14.15.1, implying inconsistently that we require that very specific version.

Done as part of https://github.com/ocaml/v3.ocaml.org/pull/332

> Make the engine version range for yarn more narrow.

I accidentally went the wrong direction and made it wider.

**Comment by rdavison** [comment]
> Be consistent with required version of node. The package.json file says 14.x, so any version in that range is allowed. That also matches what we select in Vercel (which does not allow more specific versions to be selected). However, our .nvmrc file says 14.15.1, implying inconsistently that we require that very specific version.

and

> Make the engine version range for yarn more narrow.

Moved this work to https://github.com/ocaml/v3.ocaml.org/issues/353

**Comment by rdavison** [comment]
Closing, as the build succeeds in Vercel now.  Here is a screenshot showing a Ready status. 
![Screen Shot 2021-05-20 at 2 13 19 PM](https://user-images.githubusercontent.com/630055/119028626-b0981080-b975-11eb-9a21-fa6334be3a91.png)
